File: bindings/php/src/Status.php
<?php

declare(strict_types=1);

namespace Everanium\Itb3;

/**
 * Status codes mirrored from the libitb3 C ABI
 * (cmd/cshared/internal/capi/errors.go). Numeric values are stable
 * across releases.
 */
final class Status
{
    public const OK = 0;
    public const BAD_HASH = 1;
    public const BAD_KEY_BITS = 2;
    public const BAD_HANDLE = 3;
    public const BAD_INPUT = 4;
    public const BUFFER_TOO_SMALL = 5;
    public const ENCRYPT_FAILED = 6;
    public const DECRYPT_FAILED = 7;
    public const SEED_WIDTH_MIX = 8;
    public const BAD_MAC = 9;
    public const MAC_FAILURE = 10;
    public const BLOB_MALFORMED_RECIPE = 11;
    public const RECIPE_PRIMITIVE_UNKNOWN = 12;
    public const UNKNOWN_PROFILE = 13;
    public const BLOB_MODE_MISMATCH = 19;
    public const BLOB_MALFORMED = 20;
    public const BLOB_VERSION_TOO_NEW = 21;
    public const BLOB_TOO_MANY_OPTS = 22;
    public const STREAM_TRUNCATED = 23;
    public const STREAM_AFTER_FINAL = 24;
    public const TRIPLE_CLOSED = 25;
    public const PROFILE_EXISTS = 26;
    public const INTERNAL = 99;

    private function __construct()
    {
    }
}

File: bindings/rust/src/status.rs
//! Status codes mirrored from the libitb3 C ABI
//! (`cmd/cshared/internal/capi/errors.go`). Numeric values are stable
//! across releases.

use std::fmt;

/// Integer status code returned by every libitb3 entry point.
#[repr(i32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ItbStatus {
    Ok = 0,
    BadHash = 1,
    BadKeyBits = 2,
    BadHandle = 3,
    BadInput = 4,
    BufferTooSmall = 5,
    EncryptFailed = 6,
    DecryptFailed = 7,
    SeedWidthMix = 8,
    BadMac = 9,
    MacFailure = 10,
    BlobMalformedRecipe = 11,
    RecipePrimitiveUnknown = 12,
    UnknownProfile = 13,
    Reserved14 = 14,
    Reserved15 = 15,
    Reserved16 = 16,
    Reserved17 = 17,
    BlobModeMismatch = 19,
    BlobMalformed = 20,
    BlobVersionTooNew = 21,
    BlobTooManyOpts = 22,
    StreamTruncated = 23,
    StreamAfterFinal = 24,
    TripleClosed = 25,
    ProfileExists = 26,
    Internal = 99,
}

impl TryFrom<i32> for ItbStatus {
    type Error = i32;

    fn try_from(code: i32) -> Result<Self, i32> {
        Ok(match code {
            0 => Self::Ok,
            1 => Self::BadHash,
            2 => Self::BadKeyBits,
            3 => Self::BadHandle,
            4 => Self::BadInput,
            5 => Self::BufferTooSmall,
            6 => Self::EncryptFailed,
            7 => Self::DecryptFailed,
            8 => Self::SeedWidthMix,
            9 => Self::BadMac,
            10 => Self::MacFailure,
            11 => Self::BlobMalformedRecipe,
            12 => Self::RecipePrimitiveUnknown,
            13 => Self::UnknownProfile,
            14 => Self::Reserved14,
            15 => Self::Reserved15,
            16 => Self::Reserved16,
            17 => Self::Reserved17,
            19 => Self::BlobModeMismatch,
            20 => Self::BlobMalformed,
            21 => Self::BlobVersionTooNew,
            22 => Self::BlobTooManyOpts,
            23 => Self::StreamTruncated,
            24 => Self::StreamAfterFinal,
            25 => Self::TripleClosed,
            26 => Self::ProfileExists,
            99 => Self::Internal,
            other => return Err(other),
        })
    }
}

impl fmt::Display for ItbStatus {
    /// Writes the variant's own name. The sentence describing a
    /// failure arrives already assembled in the library's diagnostic
    /// (`ItbError::Status::message`), so nothing here restates it.
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{self:?}")
    }
}

impl std::error::Error for ItbStatus {}
